src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp

```diff
@@ -0,0 +1,156 @@
+// Copyright (C) 2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "lazy_tensor.hpp"
+
+using ov::npuw::weights::ConcatMeta;
+using ov::npuw::weights::LazyTensor;
+using ov::npuw::weights::Transform;
+using ov::npuw::weights::TransformType;
+
+std::size_t LazyTensor::Hash::operator()(const LazyTensor& lt) const {
+    std::size_t seed = std::hash<void*>()(lt.m_orig_data) + 0x9e3779b9;
+    seed ^= std::hash<std::string>()(lt.m_orig_shape.to_string()) + 0x9e3779b9;
+    seed ^= std::hash<std::string>()(lt.m_orig_type.to_string()) + 0x9e3779b9;
+    for (const auto& tr : lt.m_transforms) {
+        seed ^= std::hash<int>()(static_cast<int>(tr.first)) + 0x9e3779b9;
+        if (tr.first == TransformType::PERMUTE) {
+            const auto& axes = std::get<std::vector<std::size_t>>(tr.second);
+            for (const auto& axis : axes) {
+                seed ^= std::hash<std::size_t>()(axis) + 0x9e3779b9;
+            }
+        } else if (tr.first == TransformType::CONCAT) {
+            // concat tag can be different, no need to hash it
+            const auto& axis = std::get<ConcatMeta>(tr.second).second;
+            seed ^= std::hash<std::size_t>()(axis) + 0x9e3779b9;
+        }
+    }
```

Comment:
`LazyTensor` is by design a purely functional, recursive, lazy data structure.

As your `LazyTensor` is created on top of another `LazyTensor`, all you need to do is to calc hash of your transformation + combine it with hash of your "parent" (which internally will do the same).

This hashing can and should be computed at construction time as your LazyTensor doesn't change after that.

Reply:
Done